src/main/java/com/aws/greengrass/util/platforms/windows/WindowsPlatform.java

```diff
@@ -658,6 +664,74 @@
         return "loader.cmd";
     }
 
+    @Override
+    public Map<String, Object> getOSAndKernelMetrics() {
+        return new HashMap<String, Object>() {{
+            put("OSBuildMajor", Advapi32Util.registryGetStringValue(
+                    WinReg.HKEY_LOCAL_MACHINE, REGISTRY_KEY_NAME,
+                    "CurrentBuild"
+            ));
+            put("OSBuildMinor", Advapi32Util.registryGetIntValue(
+                    WinReg.HKEY_LOCAL_MACHINE, REGISTRY_KEY_NAME,
+                    "UBR"
+            ));
+            put("Family", Advapi32Util.registryGetStringValue(
+                    WinReg.HKEY_LOCAL_MACHINE, REGISTRY_KEY_NAME,
+                    "ProductName"
+            ));
+            put("CPUArchitecture", Advapi32Util.registryGetStringValue(
+                    WinReg.HKEY_LOCAL_MACHINE, "SYSTEM\\CurrentControlSet\\Control\\Session Manager\\Environment",
+                    "PROCESSOR_ARCHITECTURE"
+            ));
+            put("OSName", getOSName());
+            put("OSVersion", getOSVersion());
+            put("KnowledgeBaseArticles", getInstalledKBs());
+        }};
+    }
+
+    private String getOSName() {
+        return executePowerShellCommand("Get-CimInstance -ClassName Win32_OperatingSystem "
+                + "| Select-Object -ExpandProperty Caption");
+    }
+
+    private String getOSVersion() {
+        return executePowerShellCommand("Get-CimInstance -ClassName Win32_OperatingSystem "
+                + "| Select-Object -ExpandProperty Version");
+    }
+
+    private String getInstalledKBs() {
+        List<String> installedKnowledgeBaseArticles = new ArrayList<>();
+        String knowledgeBaseQueryCommand = "Get-HotFix | Select-Object -ExpandProperty HotFixID";
+        try {
+            ProcessBuilder builder = new ProcessBuilder("powershell.exe",
+                    "-Command", knowledgeBaseQueryCommand);
+            Process process = builder.start();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+            String line;
+            while ((line = reader.readLine()) != null) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

An attacker can provide an input stream or file that contains an excessive number of lines without a terminating condition. The call to `readLine()` might allow an attacker to crash the program or otherwise make it unavailable to legitimate users which causes denial of service. To avoid this, make sure you either enforce a limit on the lines read or apply proper input sanitization.